Feature/ios 3987 add chia address service

skibinalexander opened the pull request:
Add integration chia blockchain

skibinalexander committed:
IOS-3987 Add implementation Chia address service

skibinalexander committed:
IOS-3987 Merge branch 'develop' into feature/IOS-3987-add-chia-address-service

skibinalexander committed:
IOS-3987 Did complete write address service

skibinalexander committed:
IOS-3987 Add implementation deserialize and decode tcxh address service chia

skibinalexander committed:
IOS-3987 Fix derivation path

Balashov152 commented on BlockchainSdk/Extensions/BlockchainUtils/Bech32.swift:
```diff
@@ -14,6 +14,9 @@ import Foundation
 /// Bech32 checksum implementation
 public class Bech32 {
+    
+    private static let BECH32M_CONST = UInt32(0x2bc830a3)
+    private static let BECH32_CONST = UInt32(1)
```
```suggestion
    private static let bech32mConst = UInt32(0x2bc830a3)
    private static let bech32Const = UInt32(1)
```
в идеале хоть какой нибудь коммент, почему так, откуда взялось

skibinalexander replied:
Ну коммент думаю не нужен, потому что это по стандарту taproot адреса, могу ссылочку закрепить, но в целом это в официальной доке в заголовке указано

Balashov152 commented on BlockchainSdk/Extensions/BlockchainUtils/Bech32.swift:
```diff
@@ -101,6 +114,7 @@ public class Bech32 {
     }
     
     /// Decode Bech32 string
+    @discardableResult
```
а зачем декодить, если не смотришь результат?)

skibinalexander replied:
> а зачем декодить, если не смотришь результат?)

Проверю, может на валидацией крутил throw

Balashov152 commented on BlockchainSdk/WalletManagers/Chia/ChiaAddressService.swift:
```diff
@@ -0,0 +1,54 @@
+    // MARK: - Implementation
+    
+    public func makeAddress(for publicKey: Wallet.PublicKey, with addressType: AddressType) throws -> PlainAddress {
+        let puzzle = Data(hex: "ff02ffff01ff02ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff0bff80808080ff80808080ff0b80ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b0") + (publicKey.blockchainKey) + Data(hex: "ff018080")
```
тоже бы документацию, почему именно такой puzzle )

skibinalexander replied:
> тоже бы документацию, почему именно такой puzzle )

Добавлю ссылку на доку

tureck1y replied:
тоже тогда надо в константы вынести, некрасиво болтается

skibinalexander replied:
> тоже тогда надо в константы вынести, некрасиво болтается

Да, окей разобью

Balashov152 commented on BlockchainSdkTests/Common/AddressTests.swift:
```diff
@@ -753,4 +753,15 @@ class AddressesTests: XCTestCase {
         XCTAssertFalse(addressService.validate("cosmos1hsk6jryyqjfhp5dhc55tc9jtckygx0eph6dd02"))
     }
+    
+    func testChiaAddressService() throws {
```
Вот тут вроде как можно поглядеть адреса, для тестов
https://github.com/irulast/chia-crypto-utils/blob/f4ab826579d19b2d2fe99d8395b3ea9b5fe8c35b/test/util/test_data.dart#L43

Balashov152 replied:
или вот тут есть разная документация
https://github.com/Chia-Network/chia-docs/blob/4f4cf5d4f07e77cef62e9cc8822ea7d632c9d593/docs/guides/crash-course/signatures.md?plain=1#L56

skibinalexander replied:
Тесты пока не финалил, это было для того чтобы пазл проверить. Добавлю в текущий МР - окс

skibinalexander committed:
IOS-3987 Fix review comments

skibinalexander committed:
IOS-3987 Fix review comments

skibinalexander committed:
IOS-3987 Merge branch 'develop' into feature/IOS-3987-add-chia-address-service

tureck1y commented on BlockchainSdk/Extensions/BlockchainUtils/Bech32.swift:
```diff
@@ -14,6 +15,9 @@ import Foundation
 /// Bech32 checksum implementation
 public class Bech32 {
+    
+    private static let bech32mConst = UInt32(0x2bc830a3)
```
Оформи пожалуйста как enum Constants в extension у Bech32.

skibinalexander replied:
> Оформи пожалуйста как enum Constants в extension у Bech32.

Да, ок, сделаю

tureck1y commented on BlockchainSdk/WalletManagers/Chia/ChiaAddressService.swift:
```diff
@@ -0,0 +1,58 @@
+        let puzzle = Data(hex: "ff02ffff01ff02ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff0bff80808080ff80808080ff0b80ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b0") + (publicKey.blockchainKey) + Data(hex: "ff018080")
+        
+        let puzzleHash = try ClvmNode.Decoder(programBytes: puzzle.bytes).deserialize().hash()
+        let encodeValue = Bech32(isBech32m: true).encode(HRP(isTestnet: isTestnet).rawValue, values: puzzleHash)
```
давай как-то облагородим. Или на переменные разнести или интерфейс надо упростить

tureck1y commented on BlockchainSdk/Extensions/BlockchainUtils/Bech32.swift:
```diff
@@ -32,6 +36,16 @@ public class Bech32 {
+    
+    // MARK: - Init
+    
+    init(isBech32m: Bool = false) {
```
что такое isBech32m?

skibinalexander replied:
> что такое isBech32m?

Формат декода сделал для 2х стандартов, bech32 это который segwit и есть новый стандарт taproot, в доке выше они 2 закреплены. Отличаются по факту константой при декодинге, нужен флаг какой, выбора константы

skibinalexander replied:
Дока Вначале файла:

//  Base32 address format for native v0-16 witness outputs implementation
//  BECH32 - https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki
//  BECH32M - https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki#L4

tureck1y commented on BlockchainSdk/WalletManagers/Chia/ChiaAddressService.swift:
```diff
@@ -0,0 +1,58 @@
+//
+
+import Foundation
+import BitcoinCore
```
можно убрать BitcoinCore?

skibinalexander replied:
> можно убрать BitcoinCore?

Да, спс, осталось после проб и ошибок

tureck1y commented on BlockchainSdk/WalletManagers/Chia/ChiaAddressService.swift:
```diff
@@ -0,0 +1,58 @@
+}
+
+extension ChiaAddressService {
+    public enum HRP: String {
```
public прям обязательно?

tureck1y commented on BlockchainSdk/Common/Utils/ClvmProgram.swift:
```diff
@@ -0,0 +1,143 @@
+}
+
+extension ClvmNode {
+    struct Iterator<T>: IteratorProtocol {
```
не получится приватным сделать?

skibinalexander replied:
> не получится приватным сделать?

Вообще хотел вынести в общий ShelIterator, если общий не нужен , Давай запривачу

skibinalexander commented on BlockchainSdk/Common/Utils/ClvmProgram.swift:
```diff
@@ -0,0 +1,143 @@
+import TangemSdk
+import CryptoSwift
+
+class ClvmNode {
```
> Если это нужно только для Chiа, то давай перенесем это в папку с Chia, не надо в Common

Вообще это универсальный механизм, поэтому вынес в common, пока используется в Chia, давай действительно пока переложу

skibinalexander commented on BlockchainSdk/Common/Utils/ClvmProgram.swift:
```diff
@@ -0,0 +1,143 @@
+import CryptoSwift
+
+class ClvmNode {
+    private(set) var atom: Array<Byte>?
+    private(set) var left: ClvmNode?
+    private(set) var right: ClvmNode?
```
Сделал приватными, на структуру не получится там рекурсия

skibinalexander committed:
IOS-3987 Fix review comments

skibinalexander committed:
IOS-3987 Merge branch 'develop' into feature/IOS-3987-add-chia-address-service

skibinalexander committed:
IOS-3987 Add need init blockchain interface

skibinalexander committed:
IOS-3987 Merge branch 'develop' into feature/IOS-3987-add-chia-address-service

Balashov152 commented on BlockchainSdk/Common/Blockchain+AllCases.swift:
```diff
@@ -94,6 +95,7 @@ extension Blockchain {
             .terraV1,
             .terraV2,
             .cronos,
+            .chia(testnet: false)
```
`.telos` пропал как будто

skibinalexander replied:
Подтянул последние изменения

Balashov152 commented on BlockchainSdk/WalletManagers/Chia/Common/ClvmProgram.swift:
```diff
@@ -0,0 +1,143 @@
+    
+    // MARK: - Hashable
+    
+    func hash() throws -> Data {
+        let hash: Data
+        
+        if let value = atom {
+            hash = Data([1] + value).sha256()
+        } else {
+            if let left = try left?.hash(), let right = try right?.hash() {
+                hash = Data([2] + left + right).sha256()
+            } else {
+                throw NSError()
+            }
+        }
+        
+        return hash
+    }
```
```suggestion
    func hash() throws -> Data {
        if let value = atom {
            return Data([1] + value).sha256()
        } else if let left = try left?.hash(), let right = try right?.hash() {
            return Data([2] + left + right).sha256()
        }
        
        throw NSError()
    }
```

Так попроще как будто)

skibinalexander replied:
> Так попроще как будто)

Поправил, дебажил, поэтому пришлось на условия разбить

Balashov152 commented on BlockchainSdk/WalletManagers/Chia/Common/ClvmProgram.swift:
```diff
@@ -0,0 +1,143 @@
+}
+
+extension ClvmNode {
+    private struct Iterator<T>: IteratorProtocol {
+        typealias Element = T
```
```suggestion
    private struct Iterator<Element>: IteratorProtocol {
```

Можно вот так сразу без `typealias`

Balashov152 commented on BlockchainSdk/WalletManagers/Chia/Common/ClvmProgram.swift:
```diff
@@ -0,0 +1,143 @@
+            return element
+        }
+        
+        mutating func next(byteCount: Int) throws -> Array<Element> {
+            var result = Array<Element>()
+            
+            for _ in 0..<byteCount {
+                guard let next = next() else {
+                    throw IteratorError.undefinedElement
+                }
+                
+                result.append(next)
+            }
+            
+            return result
+        }
+    }
```
```suggestion
    mutating func next(byteCount: Int) throws -> Array<Element> {
        (0 ..< byteCount).map {
            guard let next = next() else {
                throw IteratorError.undefinedElement
            }
            
            return next
        }
    }
```

можно вот так, ну и как есть тоже норм для дебага

Balashov152 commented on BlockchainSdkTests/Common/AddressTests.swift:
```diff
@@ -752,4 +752,15 @@ class AddressesTests: XCTestCase {
+        let blockchain = Blockchain.chia(testnet: true)
+        let addressService = ChiaAddressService(isTestnet: blockchain.isTestnet)
+        
+        let expectedAddress = try addressService.makeAddress(
+            from: Data(hex: "b8f7dd239557ff8c49d338f89ac1a258a863fa52cd0a502e3aaae4b6738ba39ac8d982215aa3fa16bc5f8cb7e44b954d")
+        ).value
```
Вообще супер если есть ссылка откуда инфа взята)
Если сам делал, то пофиг )

skibinalexander replied:
Генерили тесты с андроидом

tureck1y commented on BlockchainSdk/Extensions/BlockchainUtils/Bech32.swift:
```diff
@@ -32,6 +33,16 @@ public class Bech32 {
+    
+    // MARK: - Init
+    
+    init(constant: Constant = .bech32) {
```
Constants используем когда приватный тип. А раз в конструкторе торчит, то лучше читаемое название.  Например type, algorithm итп

skibinalexander replied:
> Constants используем когда приватный тип. А раз в конструкторе торчит, то лучше читаемое название. Например type, algorithm итп

Поменял на понятный variant

skibinalexander committed:
IOS-3987 Merge branch 'develop' into feature/IOS-3987-add-chia-address-service

# Conflicts:
#	BlockchainSdk/Common/Blockchain+AllCases.swift
#	BlockchainSdk/Common/Blockchain.swift
#	BlockchainSdk/Extensions/TWCoin+.swift

tureck1y commented on BlockchainSdk/Extensions/BlockchainUtils/Bech32.swift:
```diff
@@ -32,6 +33,16 @@ public class Bech32 {
+    
+    // MARK: - Init
+    
+    init(type: ConstantType = .bech32) {
```
Предлагаю так
`init(variant: Variant = .bech32)`
Основываясь на этом предложении из доки: 
This document defines an improved variant of Bech32 called Bech32m

skibinalexander replied:
Да , давай

skibinalexander committed:
IOS-3987 Update name bech32m variant

tureck1y commented on BlockchainSdkTests/Common/AddressTests.swift:
```diff
@@ -752,4 +752,15 @@ class AddressesTests: XCTestCase {
+            from: Data(hex: "b8f7dd239557ff8c49d338f89ac1a258a863fa52cd0a502e3aaae4b6738ba39ac8d982215aa3fa16bc5f8cb7e44b954d")
+        ).value
+        
+        XCTAssertEqual("txch14gxuvfmw2xdxqnws5agt3ma483wktd2lrzwvpj3f6jvdgkmf5gtq8g3aw3", expectedAddress)
```
нет теста на валидацию адреса, проглядел когда ревьювил в прошлый раз

skibinalexander replied:
На тесты завожу отдельную задачу всегда, на адрес, транзакцию , размер, там еще все равно жирный МР будет и изменения в тестах будут

tureck1y replied:
ну на создание есть, на валиацию тоже надо сразу. Отдельно не надо. Льем только протезированный код юнитами

skibinalexander replied:
> ну на создание есть, на валиацию тоже надо сразу. Отдельно не надо. Льем только протезированный код юнитами

Добавил по 3 теста на валидацию

tureck1y commented on BlockchainSdkTests/Common/AddressTests.swift:
```diff
@@ -752,4 +752,15 @@ class AddressesTests: XCTestCase {
+        let blockchain = Blockchain.chia(testnet: true)
+        let addressService = ChiaAddressService(isTestnet: blockchain.isTestnet)
+        
+        let expectedAddress = try addressService.makeAddress(
```
тогда это не expected же)

skibinalexander replied:
Угу тут поменяю , тестам мало внимание уделял , в следующих МР еще добавятся. Поменяю

tureck1y commented on BlockchainSdk/WalletManagers/Chia/ChiaAddressService.swift:
```diff
@@ -0,0 +1,67 @@
+}
+
+extension ChiaAddressService {
+    enum Constans: String {
```
enum Constants {
static let puzzleReveal = ....
static let  fingerprint = "ff018080"
}

skibinalexander replied:
> enum Constants {
> static let puzzleReveal = ....
> static let fingerprint = "ff018080"
> }

Вынес в Отдельный файл констант только по фингерпринтам

tureck1y commented on BlockchainSdk/WalletManagers/Chia/ChiaAddressService.swift:
```diff
@@ -0,0 +1,67 @@
+    
+    private(set) var isTestnet: Bool
+    
+    private let constans = ChiaAddressService.Constans.self
```
что-то лишнее

skibinalexander committed:
IOS-3987 Fix review comment

skibinalexander committed:
IOS-3987 Merge branch 'develop' into feature/IOS-3987-add-chia-address-service

# Conflicts:
#	BlockchainSdk/Common/Derivations/DerivationConfigV1.swift
#	BlockchainSdk/Common/Derivations/DerivationConfigV2.swift
#	BlockchainSdk/Common/Derivations/DerivationConfigV3.swift

skibinalexander committed:
IOS-3987 Update ChiaConstants

skibinalexander committed:
IOS-3987 Fix merge by develop

skibinalexander commented:
В виду того что ревью что уже в другой ветке, есть изменения с аффектом на текущие , подтащил от туда изменения + учел комменты

tureck1y commented on BlockchainSdk/Common/Address/AddressTypesConfig.swift:
```diff
@@ -52,7 +52,9 @@ struct AddressTypesConfig {
-             .octa:
+             .octa,
+             .chia
+            :
```
так оставим двоеточие?

skibinalexander replied:
После рефакторинга вроде решилось

tureck1y commented on BlockchainSdk/WalletManagers/Chia/ChiaAddressService.swift:
```diff
@@ -0,0 +1,48 @@
+    
+    private(set) var isTestnet: Bool
+    
+    private let constants = ChiaConstant.self
```
зачем?

skibinalexander replied:
Уберу, но так тоже можно же , чтобы класс не отсвечивал большими буквами, если режет глаз , напишу через обращение к классу

tureck1y replied:
это что-то совсем лишнее) Constants.constant1 ничего не режет)

skibinalexander replied:
> это что-то совсем лишнее) Constants.constant1 ничего не режет)

Урал

tureck1y commented on BlockchainSdk/WalletManagers/Chia/Common/ChiaConstant.swift:
```diff
@@ -0,0 +1,46 @@
+ - Сurried and serialized signature.clsp (https://github.com/Chia-Network/chialisp-crash-course/blob/af620db2505db507b348d4f036dc4955fa81a004/signature.clsp)
+ */
+
+enum ChiaConstant {
```
- не надо все константы со всех файлов собирать в одном месте
- я же написал уже паттерн, посмотри пожалуйста как мы делаем с enum Constants
- методам тут не место, пусть будут только там, где вызываются

skibinalexander replied:
Эти методы нужны будут как в адрес сервисе , так в билдере транзакций , можно просто переименовать как-то ChiaHelpers?

skibinalexander replied:
Нужно где - то хранить статичные элементы , название константы окей не подходит , надо другое , либо предложение как разнести ? Константы отдельно методы дублировать по сущностям ?

skibinalexander replied:
И для наеймспейса говорили enum ок использовать

tureck1y replied:
Если переиспользуется, то надо cделать какую-нибудь ChiaUtils и там все сделать

skibinalexander replied:
> Если переиспользуется, то надо cделать какую-нибудь ChiaUtils и там все сделать

Ага разнес по файлам, и сделал ChiaPuzzle сущность

tureck1y replied:
Так этот то файл зачем остался, его надо спрятать в ChiaPuzzle. Посмотри пожалуйста как константы оформлены в BLSUtils. Надо сделать также

skibinalexander committed:
IOS-3987 Merge branch 'develop' into feature/IOS-3987-add-chia-address-service

skibinalexander committed:
IOS-3987 Fix review comments

skibinalexander committed:
IOS-3987 Add fix verify implementation Bech32m

skibinalexander committed:
IOS-3987 Update types config

skibinalexander committed:
IOS-3987 Fix review comments after merge develop

tureck1y commented on BlockchainSdk/WalletManagers/Chia/Common/ChiaGenesisChallenge.swift:
```diff
@@ -0,0 +1,18 @@
+
+import Foundation
+
+enum ChiaGenesisChallenge {
```
нигде не используется, поэтому я не могу понять в каком виде эта структура должна существовать, давай добавим когда уже будет использование

skibinalexander replied:
> нигде не используется, поэтому я не могу понять в каком виде эта структура должна существовать, давай добавим когда уже будет использование

Убрал

tureck1y commented on BlockchainSdk/WalletManagers/Chia/Common/ChiaHRP.swift:
```diff
@@ -0,0 +1,18 @@
+import Foundation
+
+/// Prefix address Chia blockchain
+enum ChiaHRP: String {
```
1) используется только в адрес сервисе? Тогда пусть внутри и будет.
2) Странная конструкция enum c init. Даже просто статик функция будет компактнее выглядеть

tureck1y commented on BlockchainSdk/WalletManagers/Chia/Common/ChiaPuzzle.swift:
```diff
@@ -0,0 +1,22 @@
+    static func getPuzzleHash(address: String) throws -> Data {
+        let bech32 = Bech32(variant: .bech32m)
+        let dataBytes = try bech32.decode(address).checksum
+        return try Data(bech32.convertBits(data: dataBytes.bytes, fromBits: 5, toBits: 8, pad: false))
```
вместо Data(hex: лучше использовать Extensions из TangemSdk

skibinalexander replied:
> вместо Data(hex: лучше использовать Extensions из TangemSdk

- Переделал

tureck1y commented on BlockchainSdk/WalletManagers/Chia/Common/ChiaPuzzle.swift:
```diff
@@ -0,0 +1,22 @@
+
+// PuzzleHash Chia documentation - https://docs.chia.net/guides/crash-course/signatures/
+struct ChiaPuzzle {
+    static func getPuzzle(walletPublicKey: Data) -> Data {
```
может лучше
getPuzzle(from/for walletPublicKey: Data)
getPuzzleHash(from/for address: String)

Я бы еще статику убрал

skibinalexander committed:
IOS-3987 Fix review comments

skibinalexander committed:
IOS-3987 Merge branch 'develop' into feature/IOS-3987-add-chia-address-service
